libs/@local/harpc/codec/src/error.rs

```diff
@@ -0,0 +1,299 @@
+            assert!(ones != 0, "value must not be 0");
+        }
+
+        #[expect(unsafe_code, reason = "checked that it is never 0")]
+        // SAFETY: $value is not 0
+        unsafe {
+            NonZero::new_unchecked($n)
+        }
+    }};
+}
+
+// we use a macro here to define the error codes, as the code is quite repetetive and also error
+// prone, we might not be able to increment values correctly, another problem is that rustfmt will
+// reorder the constants, making keeping tracks of the ids harder than it should be.
+macro_rules! define {
+    ($($base:literal => [$($name:ident),+]),*) => {
+        $(
+            impl ErrorCode {
+                $(
+                    pub const $name: Self = Self(non_zero!(($base as u16) + ${index(0)}));
+                )+
+            }
+        )*
+    };
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+#[cfg_attr(feature = "proptest", derive(test_strategy::Arbitrary))]
+pub struct ErrorCode(NonZero<u16>);
+
+impl ErrorCode {
+    #[must_use]
+    pub const fn new(value: NonZero<u16>) -> Self {
+        Self(value)
+    }
+
+    #[must_use]
+    pub const fn value(self) -> NonZero<u16> {
+        self.0
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for ErrorCode {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = u16::deserialize(deserializer)?;
+
+        NonZero::new(value)
+            .map(Self)
+            .ok_or_else(|| serde::de::Error::custom("value must not be 0"))
+    }
+}
+
+define! {
+    // 0xFE_xx = client errors
+    // Initiated by client, but occur on server (tower level)
+    0xFE_10 => [
+        NOT_FOUND // akin to 404
+    ],
+    // 0xFF_xx = server errors
+    // Server Session Errors
+    0xFF_00 => [
+        CONNECTION_CLOSED,
+        CONNECTION_SHUTDOWN,
+        CONNECTION_TRANSACTION_LIMIT_REACHED,
+        INSTANCE_TRANSACTION_LIMIT_REACHED,
+        TRANSACTION_LAGGING
+    ],
+    // Server Tower Errors
+    0xFF_10 => [
+        PACK_INVALID_ERROR_TAG
+    ],
+    // Generic Server Errors
+    0xFF_F0 => [
+        INTERNAL_SERVER_ERROR
+    ]
+}
+
+/// An error that is has been fully encoded and can be sent or received over the network.
+///
+/// Essentially a compiled version of a `NetworkError` or `Report<C>` into it's wire format.
+///
+/// An `EncodedError` is constructed through the `ErrorBuffer`.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct EncodedError {
+    code: ErrorCode,
+    bytes: Bytes,
+}
+
+impl EncodedError {
+    pub fn new(code: ErrorCode, bytes: Bytes) -> Option<Self> {
+        let &first = bytes.first()?;
+
+        kind::Tag::variants()
+            .into_iter()
+            .any(|tag| tag as u8 == first)
+            .then(|| Self { code, bytes })
+    }
```

Comment:
kinda, added this specifically for the upper `harpc-tower` layer, the Response `Body` is a stream of `ResponseKind` and `Buf`, so it could be that `Response::INTERNAL_SERVER_ERROR`, followed by `Bytes::from_static("hello world")`, that's invalid, so would result in an invalid `EncodedError`, returning `None` here allows us to not panic, but instead still return an error to the client indicating that the error response was malformed (which it should never be).

So I mean I get either 🤔.

Reply:
I just prioritized getting a response to the client, instead of crashing the server